### PR TITLE
Implement new abstract lephare functions

### LINKED ISSUE #24: Implement new abstracted lephare functions
We are in the process of creating abstract prepare, and process functions which should abstract away a lot of the logic.

I also notice that currently estimate_pdf is defined twice in the estimate stage. This needs to be fixed alongside the update.

### LINKED ISSUE #30: Fix hardcode magnitude column names
Hello, when running an estimation with a file using different column names for the magnitudes, I get the following error:

```
Process 0 running estimator on chunk 0 - 1561
Traceback (most recent call last):
  File "/lustre/t1/cl/lsst/tmp/henrique.almeida/slurm-home/bin/rail-estimate", line 260, in <module>
    main()
  File "/lustre/t1/cl/lsst/tmp/henrique.almeida/slurm-home/bin/rail-estimate", line 256, in main
    estimate(cfg, ctx)
  File "/lustre/t1/cl/lsst/tmp/henrique.almeida/slurm-home/bin/rail-estimate", line 248, in estimate
    ctx.estimator.estimate(ctx.input)
  File "/lustre/t1/cl/lsst/tmp/henrique.almeida/miniconda3/lib/python3.11/site-packages/rail/estimation/estimator.py", line 97, in estimate
    self.run()
  File "/lustre/t1/cl/lsst/tmp/henrique.almeida/miniconda3/lib/python3.11/site-packages/rail/estimation/estimator.py", line 110, in run
    self._process_chunk(s, e, test_data, first)
  File "/lustre/t1/cl/lsst/tmp/henrique.almeida/miniconda3/lib/python3.11/site-packages/rail/estimation/algos/lephare.py", line 214, in _process_chunk
    flux = np.array([data["mag_{}_lsst".format(b)] for b in "ugrizy"]).T
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/lustre/t1/cl/lsst/tmp/henrique.almeida/miniconda3/lib/python3.11/site-packages/rail/estimation/algos/lephare.py", line 214, in <listcomp>
    flux = np.array([data["mag_{}_lsst".format(b)] for b in "ugrizy"]).T
                     ~~~~^^^^^^^^^^^^^^^^^^^^^^^^^
KeyError: 'mag_u_lsst'
```

I'm passing with the common parameters the correct name ("mag_u").

- [x] I have described the situation in which the bug arose, including what code was executed, information about my environment, and any applicable data others will need to reproduce the problem.
- [x] I have included available evidence of the unexpected behavior (including error messages, screenshots, and/or plots) as well as a descriprion of what I expected instead.
- [x] If I have a solution in mind, I have provided an explanation and/or pseudocode and/or task list.

### PR DESCRIPTION
**Pull request by raphaelshirley**
we are trying to move most of the lephare code to lephare itself.

Closes #24 
Closes #30 

This will also involve changes the Params to take lephare configs to overwrite values

Do we want to run get data in the info stages?

What minimal functionality do we want to have in place prior to the in person meetings?

We would like the option to run autoadapt either on the inform training set or on each chunk

## Problem & Solution Description (including issue #)

rail_lephare is running but we decided it would be neater to move some of the code to lephare in abstract functions rather than have details exposed to rail_lephare. this involves two new functions prepare and process in lephare which should be the totality of what would be run in rail_lephare. 

## Code Quality
- [ ] My code follows [the code style of this project](https://rail-hub.readthedocs.io/en/latest/source/contributing.html#naming-conventions)
- [ ] I have written unit tests or justified all instances of `#pragma: no cover`; in the case of a bugfix, a new test that breaks as a result of the bug has been added
- [ ] My code contains relevant comments and necessary documentation for future maintainers; the change is reflected in applicable demos/tutorials (with output cleared!) and added/updated docstrings use the [NumPy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html)
- [ ] Any breaking changes, described above, are accompanied by backwards compatibility and deprecation warnings

### CODE REVIEW

**Comment by raphaelshirley** [comment]
We might want to run the get data which would also close #12 

Further we should add the autoadapt stage to the inform stage. In general we will not have access to spectroscopic redshifts in the estimate stage.

**Comment by raphaelshirley** [comment]
I left get data out of the inform stage so it must be run outside of rail.

**Comment by raphaelshirley** [comment]
I think the tests are failing because the lephare version doesn't yet have the change to create_new_run which we now use to set the run equal to the informer name.

**Comment by drewoldag** [comment]
Yep, that makes sense.

File: src/rail/estimation/algos/lephare.py [hunk 1]
```diff
@@ -25,63 +27,51 @@ class LephareInformer(CatInformer):
...
-            "$LEPHAREDIR/examples/COSMOS_MOD.list",
-            msg="Path to text file containing list of quasar SED templates",
-        ),
-        star_mag_dict=Param(
+        lephare_config=Param(
             dict,
-            dict(
-                lib_ascii="YES",
+            lp.read_config(
+                "{}/{}".format(os.path.dirname(os.path.abspath(__file__)), "lsst.para")
```
Comment (drewoldag) [inline src/rail/estimation/algos/lephare.py:33]: I suppose this is fine - but I don't think I've seen the default value being defined as a line of code like this. i.e. lp.read_config(...). The way this was originally seemed more intuitive. With `self.lephare_config = lp.read_config(self.config["lephare_config_file"])` specified in the `__init__` method.

File: src/rail/estimation/algos/lephare.py [hunk 2]
```diff
@@ -206,52 +196,104 @@ def _process_chunk(self, start, end, data, first):
...
+    use different runs.
+
+    This simple function updates the environment variables and reloads lephare
+    to ensure they are properly used.
+    """
+    if lepharedir is not None:
+        os.environ["LEPHAREDIR"] = lepharedir
+    if lepharework is not None:
+        os.environ["LEPHAREWORK"] = lepharework
+    importlib.reload(lp)
```
Comment (drewoldag) [inline src/rail/estimation/algos/lephare.py:279]: If this is working, then that's fine, I don't think it's worth poking a bear. But I really do wish we had a way around this that didn't require it.

**Review by drewoldag — APPROVED** [review]
Overall I think this looks just fine. I'm sure there are corner cases and issues that we haven't uncovered yet, but I would advocate for moving to get it in to main so that we can all experiment with it together.

**Pull request merged**